server/gfmd/user.c

```diff
@@ -2011,18 +2117,49 @@ gfm_server_user_auth_modify(struct peer *peer,
 	if (e == GFARM_ERR_NO_ERROR) {
 
 		if (auth_user_id == NULL || strcmp(auth_user_id, "") == 0) {
-			e = user_auth_id_remove(u, auth_user_id_type_str);
+			bool need_to_update_db = false;
+			e = user_auth_id_remove_internal(u,
+				auth_user_id_type_str, &need_to_update_db);
+
+			if (e == GFARM_ERR_NO_ERROR &&
+				need_to_update_db) {
+
+				struct db_user_auth_remove_arg arg = {
+					username,
+					auth_user_id_type_str
+				};
+				e = db_user_auth_remove(&arg);
+			}
 		} else {
-			e = user_auth_id_modify(u,
-				auth_user_id_type_str, auth_user_id);
+			bool need_to_update_db = false;
+			bool need_to_add = false;
+			e = user_auth_id_modify_internal(u,
+				auth_user_id_type_str, auth_user_id,
+				&need_to_update_db, &need_to_add);
+
+			if (e == GFARM_ERR_NO_ERROR &&
+				need_to_update_db) {
+
+				struct db_user_auth_arg arg = {
+					username,
+					auth_user_id_type_str,
+					auth_user_id
+				};
+
+				if (need_to_add) {
+					e = db_user_auth_add(&arg);
+				} else {
+					e = db_user_auth_modify(&arg);
+				}
```

Comment:
the error of db_user_auth_add/modify() should be logged by gflog_error(), it's somewhat critical error.